### PR TITLE
rewire SaveRecurringTip through ledger

### PR DESCRIPTION
**Pull request by cg505**
Resolves brave/brave-browser#5152.

@kylehickinson These iOS changes build, but I didn't actually build them into a full build. If it's straightforward to make sure this doesn't break anything, that would be good to test.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [x] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
1. Enable rewards and go to a verified publisher.
1. Add a montly tip for the publisher.
1. Verify that you see the montly contribution in `brave://rewards`.
1. To back the the publisher page and open the wallet panel. Change the tip amount.
1. Verify that the new tip amount is show in `brave://rewards`.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 1]
```diff
@@ -466,19 +466,20 @@ - (void)listRecurringTips:(void (NS_NOESCAPE ^)(NSArray<BATPublisherInfo *> *))c
 
 - (void)addRecurringTipToPublisherWithId:(NSString *)publisherId amount:(double)amount
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: We should add a completion block here that is called in the `SaveRecurringTip`.  suggested declaration:

`- (void)addRecurringTipToPublisherWithId:(NSString *)publisherId amount:(double)amount completion:(void (^)(BOOL success))completion`

Needs to updated in header as well

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 2]
```diff
@@ -466,19 +466,20 @@ - (void)listRecurringTips:(void (NS_NOESCAPE ^)(NSArray<BATPublisherInfo *> *))c
...
-                                                          observer.recurringTipAdded(publisherId);
-                                                        }
-                                                      }
-                                                    }];
+  ledger::ContributionInfoPtr info = ledger::ContributionInfo::New();
+  info->publisher = publisherId.UTF8String;
+  info->value = amount;
+  info->date = [[NSDate date] timeIntervalSince1970];
+  ledger->SaveRecurringTip(std::move(info), ^(ledger::Result result){
+    if (result != ledger::Result::LEDGER_OK) {
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: completion block should be called above this line after adding it (seen in previous comment): 
```objc
dispatch_async(dispatch_get_main_queue(), ^{
  completion(result == ledger::Result::LEDGER_OK)
});
```

Comment (kylehickinson) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: edit: you don't need the `dispatch_async` to main queue, the database insertion callbacks are done on main queue already

**Comment by cg505** [comment]
tests pass, some CI issues only

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 3]
```diff
@@ -2213,25 +2213,64 @@ bool SaveRecurringTipOnFileTaskRunner(
...
+      std::move(info),
+      base::BindOnce(&RewardsServiceImpl::OnSaveRecurringTipUI,
+                     AsWeakPtr(),
+                     std::move(callback)));
+}
+
+void RewardsServiceImpl::OnRecurringTipSaved(
+    ledger::SaveRecurringTipCallback callback,
+    bool success) {
   for (auto& observer : observers_) {
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: this observers should be in `OnSaveRecurringTipUI`

File: components/brave_rewards/browser/rewards_service_impl.h [hunk 4]
```diff
@@ -210,8 +210,13 @@ class RewardsServiceImpl : public RewardsService,
       const std::string& publisher_key,
       RefreshPublisherCallback callback) override;
 
-  void SaveRecurringTip(const std::string& publisher_key,
-                             const int amount) override;
+  void OnSaveRecurringTipUI(
+      SaveRecurringTipCallback callback,
+      int32_t result);
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.h]: nit: const

File: components/brave_rewards/browser/rewards_service_impl.h [hunk 5]
```diff
@@ -341,7 +346,9 @@ class RewardsServiceImpl : public RewardsService,
              ledger::PublisherInfoPtr publisher_info);
   void OnContributionInfoSaved(const ledger::REWARDS_CATEGORY category,
                                bool success);
-  void OnRecurringTipSaved(bool success);
+  void OnRecurringTipSaved(
+      ledger::SaveRecurringTipCallback callback,
+      bool success);
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.h]: nit: const

**Comment by cg505** [comment]
CI passes everything but Windows here: https://staging.ci.brave.com/job/brave-browser-build-pr/job/save-recurring-tip-refactor/9//flowGraphTable/ (windows failed due to a CI bug)

**Comment by cg505** [comment]
Windows CI pass for reference: https://staging.ci.brave.com/job/brave-browser-build-pr/job/save-recurring-tip-refactor/11//flowGraphTable/

**Pull request merged**